src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs

```diff
@@ -0,0 +1,979 @@
+        }
+
+        // SQLBUVSTS01 bugs 192677 and 193221: we cannot use System.IO.Path.GetFullPath for two reasons:
+        // * it requires PathDiscovery permissions, which is unnecessary for SqlFileStream since we are dealing with network path
+        // * it is limited to 260 length while in our case file path can be much longer
+        // To overcome the above limitations we decided to use GetFullPathName function from kernel32.dll
+        [ResourceExposure(ResourceScope.Machine)]
+        [ResourceConsumption(ResourceScope.Machine)]
+        static private string GetFullPathInternal(string path)
+        {
+            //-----------------------------------------------------------------
+            // precondition validation should be validated by callers of this method
+            // NOTE: if this method moves elsewhere, this assert should become an actual runtime check
+            // as the implicit assumptions here cannot be relied upon in an inter-class context
+            Debug.Assert(path != null);
+
+            // remove leading and trailing whitespace
+            path = path.Trim();
+            if (path.Length == 0)
+            {
+                throw ADP.Argument(SR.GetString(SR.SqlFileStream_InvalidPath), "path");
+            }
+
+            // check for the path length before we normalize it with GetFullPathName
+            if (path.Length > MaxWin32PathLength)
+            {
+                // cannot use PathTooLongException here since our length limit is 32K while
+                // PathTooLongException error message states that the path should be limited to 260
+                throw ADP.Argument(SR.GetString(SR.SqlFileStream_InvalidPath), "path");
+            }
+
+            // GetFullPathName does not check for invalid characters so we still have to validate them before
+            if (path.IndexOfAny(s_invalidPathChars) >= 0)
+            {
+                throw ADP.Argument(SR.GetString(SR.SqlFileStream_InvalidPath), "path");
+            }
+
+            // make sure path is a UNC path
+            if (!path.StartsWith(@"\\", StringComparison.OrdinalIgnoreCase))
+            {
+                throw ADP.Argument(SR.GetString(SR.SqlFileStream_InvalidPath), "path");
+            }
+
+            //-----------------------------------------------------------------
+
+            // normalize the path
+            path = SafeGetFullPathName(path);
+
+            // we do not expect windows API to return invalid paths
+            Debug.Assert(path.Length <= MaxWin32PathLength, "GetFullPathName returns path longer than max expected!");
+
+            // CONSIDER: is this a precondition validation that can be done above? Or must the path be normalized first?
+            // after normalization, we have to ensure that the path does not attempt to refer to a root device, etc.
+            if (path.StartsWith(@"\\.\", StringComparison.Ordinal))
```

Comment:
This isn't correct. `\\?\` is the same as `\\.\`, except it gets around MaxPath checks in Win32 and avoids normalization. If you're checking for `IsUnc` (as described above) _after_ normalizing you can be positive that you have a UNC, not some other device path. Move the IsUnc() check to this position and you'll have the intended behavior.